bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryMemTableWithParallelFlusher.java

```diff
@@ -0,0 +1,142 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.bookkeeper.bookie;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.NavigableSet;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.bookkeeper.bookie.Bookie.NoLedgerException;
+import org.apache.bookkeeper.bookie.CheckpointSource.Checkpoint;
+import org.apache.bookkeeper.common.util.OrderedExecutor;
+import org.apache.bookkeeper.conf.ServerConfiguration;
+import org.apache.bookkeeper.stats.StatsLogger;
+import org.apache.bookkeeper.util.SafeRunnable;
+
+/**
+ * EntryMemTableWithParallelFlusher.
+ */
+@Slf4j
+public class EntryMemTableWithParallelFlusher extends EntryMemTable {
+
+    final OrderedExecutor flushExecutor;
+    private static final int MEMTABLE_FLUSH_TIMEOUT_IN_SECONDS = 60;
+
+    public EntryMemTableWithParallelFlusher(final ServerConfiguration conf, final CheckpointSource source,
+            final StatsLogger statsLogger, OrderedExecutor flushExecutor) {
+        super(conf, source, statsLogger);
+        this.flushExecutor = flushExecutor;
+    }
+
+    /**
+     * Functionally this overridden flushSnapshot does the same as
+     * EntryMemTable's flushSnapshot, but it uses flushExecutor
+     * (OrderedExecutor) to process an entry through flusher.
+     *
+     * <p>entryKeyValues of a ledger are grouped in a list and runnable is created
+     * to flush process all those entries and that runnable is submitted to the
+     * flushExecutor with ledgerId as the orderingKey.
+     */
+    @Override
+    long flushSnapshot(final SkipListFlusher flusher, Checkpoint checkpoint) throws IOException {
+        AtomicLong flushedSize = new AtomicLong();
+        if (this.snapshot.compareTo(checkpoint) < 0) {
+            synchronized (this) {
+                EntrySkipList keyValues = this.snapshot;
+                if (keyValues.compareTo(checkpoint) < 0) {
+                    NavigableSet<EntryKey> keyValuesSet = keyValues.keySet();
+                    Map<Long, List<EntryKeyValue>> entryKeyValuesMap = new HashMap<Long, List<EntryKeyValue>>();
```

Comment:
nit: have you benchmarked this before? my feeling this is going to be a gc churning ...

Reply:
changed the logic here. with the new implementation there wont be any need of auxiliary datastructures in this method implementation.